pkg/container/docker_run.go

```diff
@@ -599,69 +604,98 @@ func (cr *containerReference) copyDir(dstPath string, srcPath string, useGitIgno
 			ignorer = gitignore.NewMatcher(ps)
 		}
 
-		err = filepath.Walk(srcPath, func(file string, fi os.FileInfo, err error) error {
-			if err != nil {
-				return err
+		var getCallback func(submodulePath []string) func(file string, fi fs.FileInfo, err error) error
+		getCallback = func(submodulePath []string) func(file string, fi os.FileInfo, err error) error {
```

Comment:
This function does not use `cr *containerReference`, so for readability and testability you could extract it into its own function outside of `copyDir`.
And for the test maybe it is worth looking into https://pkg.go.dev/testing/fstest to create an in-memory FS. We're doing something like that for the artifact server tests: https://github.com/nektos/act/blob/master/pkg/artifacts/server_test.go#L24-L52

Reply:
Good idea to extract it in it's own function, also may extract the tar usage into an interface to be reusable for a non docker copydir. ( run on host )

But how would I checkout a git repository in fstest? Only billyfs works with go-git if I'm not mistaken.

Reply:
Ah, I haven't thought about the "how" too much but just remembered that fs stuff can be mocked with an in-memory fs.
Maybe it'd be worth to look into go-git to see how they test things?
Or if a mocked fs doesn't work you could probably create a temp directory and test in there?

Btw, with this change it is possible to run the `checks.yml` test to test act itself, which previously failed because of `.env` and `.secrets` files 🎉 
```
$ ./dist/local/act -W ./.github/workflows/checks.yml pull_request -j test-linux
```

Reply:
> Btw, with this change it is possible to run the checks.yml test to test act itself

No idea if this still works :), after adding filesystem and output interfaces. _I don't want to test that myself._

I added an in memory test with go-billy and go-git.
Covers the two fixed bugs, no submodules.